src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java

```diff
@@ -170,13 +179,65 @@ public String addImport(String namespace) {
     return csharpCommon.addImport(namespace);
   }
 
-  private String packageNameAndImport(String suffix) {
-    String packageName =
-        CSharpContextCommon.s_underscoresToPascalCase(getApiaryConfig().getServiceCanonicalName());
-    String versionName = getApiaryConfig().getServiceVersion().replace('.', '_');
-    String namespace = "Google.Apis." + packageName + "." + versionName + suffix;
-    addImport(namespace);
-    return packageName;
+  // Put any using aliases at the end of other usings, with a line gap.
+  public Iterable<String> sortUsings(Iterable<String> usings) {
+    Predicate<String> isAlias =
+        new Predicate<String>() {
+          @Override
+          public boolean apply(String using) {
+            return using.contains("=");
+          }
+        };
+    FluentIterable<String> fluentUsings = FluentIterable.from(usings);
+    Iterable<String> aliases =
+        fluentUsings.anyMatch(isAlias)
+            ? FluentIterable.of(new String[] {""}).append(fluentUsings.filter(isAlias))
+            : Collections.<String>emptyList();
+    return fluentUsings
+        .filter(Predicates.<String>not(isAlias))
+        .append(aliases)
+        .transform(
+            new Function<String, String>() {
+              @Override
+              public String apply(String using) {
+                return using == "" ? "" : ("using " + using + ";");
+              }
+            });
+  }
+
+  private String using(String fullTypeName) {
+    if (fullTypeName.startsWith(serviceNamespace)) {
+      // Special handling of nested types within the service class, except .Data classes.
+      // Return the full nested name to place in generated source code.
+      if (fullTypeName.startsWith(serviceNamespace + ".Data")) {
+        // Add an alias for Data, required due to class name clashes in some libraries.
+        addImport("Data = " + serviceNamespace + ".Data");
+      }
+      addImport(serviceNamespace);
+      return fullTypeName.substring(serviceNamespace.length() + 1);
+    }
+    // Remove any generic information, if present.
+    int genericIndex = fullTypeName.indexOf('<');
+    String nonGenericTypeName;
+    if (genericIndex >= 0) {
+      nonGenericTypeName = fullTypeName.substring(0, genericIndex);
+    } else {
+      nonGenericTypeName = fullTypeName;
+    }
+    int typeIndex = nonGenericTypeName.lastIndexOf('.');
+    if (typeIndex < 0) {
+      throw new IllegalArgumentException("fullTypeName must be fully specified.");
+    }
+    addImport(nonGenericTypeName.substring(0, typeIndex));
+    return fullTypeName.substring(typeIndex + 1);
+  }
+
+  private String usingData(String dataRelativeTypeName) {
```

Comment:
I'd personally use a conditional operator here, but it's definitely a matter of personal preference.
(Likewise at line 222)